crates/bevy_ecs/src/storage/sorted_small_vec.rs

```diff
@@ -0,0 +1,259 @@
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct SortedSmallVec<const N: usize>(SmallVec<[usize; N]>);
+
+impl<const N: usize> IntoIterator for SortedSmallVec<N> {
+    type Item = usize;
+    type IntoIter = <SmallVec<[usize; N]> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<const N: usize> Default for SortedSmallVec<N> {
+    fn default() -> Self {
+        Self::new_const()
+    }
+}
+
+impl<const N: usize> SortedSmallVec<N> {
+    /// Construct an empty vector
+    pub fn new() -> Self {
+        Self(SmallVec::new())
+    }
+
+    /// Construct an empty vector
+    ///
+    /// This is a `const` version of [`SortedSmallVec::new()`]
+    pub(crate) const fn new_const() -> Self {
+        Self(SmallVec::new_const())
+    }
+
+    /// Construct a new `SortedSmallVec` from a `Vec<usize>`.
+    ///
+    /// Elements are copied and put in a sorted order if the original `Vec` isn't ordered.
+    /// Duplicates are removed.
+    pub fn from_vec(vec: Vec<usize>) -> Self {
+        let mut sorted_vec = Self(SmallVec::with_capacity(vec.len()));
+        for value in vec {
+            sorted_vec.insert(value);
+        }
+        sorted_vec
+    }
+
+    pub(crate) fn ones(&self) -> impl Iterator<Item = usize> + '_ {
+        self.0.iter().copied()
+    }
+
+    /// Insert the value if it's not already present in the vector.
+    /// Maintains a sorted order.
+    pub(crate) fn insert(&mut self, index: usize) {
+        match self.0.binary_search(&index) {
+            // element already present in the vector
+            Ok(_) => {}
+            Err(pos) => {
+                self.0.insert(pos, index);
+            }
+        }
+    }
+
+    /// Removes a value if it's present in the vector
+    pub(crate) fn remove(&mut self, index: usize) {
+        if let Ok(pos) = self.0.binary_search(&index) {
+            self.0.remove(pos);
+        }
+   }
+
+    /// Returns true if the vector contains the value.
+    pub(crate) fn contains(&self, index: usize) -> bool {
+        self.0.binary_search(&index).is_ok()
+    }
+
+    /// Returns true if the vector is empty.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Empties the contents of the vector
+    pub(crate) fn clear(&mut self) {
+        self.0.clear();
+    }
+
+    /// Returns the number of elements in the vector.
+    pub(crate) fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Adds all the elements from `other` into this vector. (skipping duplicates)
+    pub(crate) fn union_with(&mut self, other: &Self) {
+        let mut i = 0;
+        let mut j = 0;
+        while i < self.len() && j < other.len() {
+            match self.0[i].cmp(&other.0[j]) {
+                Ordering::Less => i += 1,
+                Ordering::Greater => {
+                    self.0.insert(i, other.0[j]);
```

Comment:
While I'm looking at this, I think I got the bit at the end wrong.  `split_at_mut` won't help because the slices would overlap.  Maybe it should be

```rust
} // end of while loop
// Shift any remaining elements from self into place
// This also removes any extra space
self.0.drain(i..w);
// Copy any remaining elements from other into place
// `drain` doesn't shrink the allocation, so this won't need to allocate
self.0.extend_from_slice(&other.0[j..]);
```

Reply:
I actually noticed that all of the 'extend' calls could practically be replaced by 'union_with' as most of them came in the form `set.extend(other_set.iter())`